perun/collect/trace/strategy.py

```diff
@@ -28,213 +28,287 @@ class _Status(IntEnum):
+        <name>, <name>begin, <name>entry,  <name>start,  <name>create,  <name>construct
+                <name>end,   <name>return, <name>finish, <name>destroy, <name>deconstruct.
+     - the pairing algorithm first tries to pair the exact combination (e.g. begin, end) and if
+       such exact combination is not found, then tries to pair rules among other combinations
 
-     :param list probes: the list of probes (as dicts) that should be paired
-     :returns list: probe dictionaries with optionally added 'pair' key containing paired probe name
-    """
-    result = []
-    for probe in probes:
-        # Split the probe definition into pair or probes
-        delim = probe['name'].find('#')
-        if delim != -1:
-            probe['pair'] = probe['name'][delim + 1:]
-            probe['name'] = probe['name'][:delim]
-            result.append(probe)
-        elif not probe['name'].endswith('_end') and not probe['name'].endswith('_END'):
-            # Find the pair probe automatically as <name>_end template
-            pair = next((pair_probe for pair_probe in probes if
-                         (pair_probe['name'] == probe['name'] + '_end' or
-                          pair_probe['name'] == probe['name'] + '_END')), None)
-            if pair:
-                probe['pair'] = pair['name']
-            result.append(probe)
-    return result
-
-
-def _merge_probes_lists(probes, probes_sampled, global_sampling):
-    """Merges the probe lists without and with specified sampling into one list with
-    unified sampling specification
+     :param dict probes: the list of probes (as dicts) that should be paired
 
-    :param list probes: list of strings that represent probe names
-    :param list probes_sampled: list of tuples that contain 0) probe name, 1) probe sampling
-    :param global_sampling: the global sampling value that is applied to all probes without sampling
-    :return list: list of probes in unified format (dictionaries with 'name' and 'sample' keys)
+     :return dict: probe dictionaries with optionally added 'pair' key containing paired probe name
     """
-    # Add global sampling (default 0) to the probes without sampling specification
-    probes = [{'name': probe, 'sample': global_sampling} for probe in probes]
-
-    # Validate the sampling values and merge the lists
-    for probe in probes_sampled:
-        if probe[1] > 1:
-            probes.append({'name': probe[0], 'sample': probe[1]})
+    result = dict()
+
+    # Direct suffix pairs that should be paired
+    suffix_pairs = {'begin': 'end', 'entry': 'return', 'start': 'finish',
+                    'create': 'destroy', 'construct': 'deconstruct'}
+    # Related suffixes that could still be paired together
+    related = (['end', 'return', 'finish'], ['destroy', 'deconstruct'])
+
+    delimited, beginnings, endings = _classify_static_probes(probes, suffix_pairs)
+
+    # First process the delimited probes
+    for k, v in delimited.items():
```

Comment:
Fix: Extract v[0] to something meaningful as well (you are already extracting v[1]). Lines 160-161 will be better readable as well.